### PR TITLE
Impl PluginMessage packet.

### PR DESCRIPTION
**Pull request by toqueteos**
Moved `src/packet.rs` to `src/packet/mod.rs` (also reordered imports).

`PluginMessage` works for both `0x3f` (clientbound) and `0x17` (serverbound).

It does add one restriction to `packet::Protocol`, `proto_decode`'s `Read` must be a Cursor, not the actual `TcpStream` (or `BufStream`). This should be easily handled by our `PacketHandler` (or whatever is called), once we have it.

/cc @fenhl @eddyb

### CODE REVIEW

**Comment by fenhl** [comment]
I'm currently refactoring `packets!` and adding a `Packet` trait whose `decode` takes the packet length. If we can land that first, we don't have to use the `Cursor` hack for this.

**Comment by toqueteos** [comment]
Hold by @fenhl proposal, old draft: https://www.irccloud.com/pastebin/HuOuHXvg/packet.rs

**Comment by fenhl** [comment]
New draft: #24

**Comment by toqueteos** [comment]
Updated for #24.

File: src/packet.rs [hunk 1]
```diff
@@ -451,7 +451,21 @@ packets! {
             // 0x3c => UpdateScore { score_name: String, action: ScoreAction }
             0x3d => DisplayScoreboard { position: i8, score_name: String }
             // 0x3e => UpdateTeam { team_name: String, action: TeamAction }
-            // 0x3f => PluginMessage { channel: String, data: Vec<u8>; encode { ... }; decode { ... }; } // PROBLEM: length of `data` comes from packet length
+            0x3f => PluginMessage { channel: String, data: Vec<u8>;
+                encode {
+                    try!(<String as Protocol>::proto_encode(self.channel, dst));
+                    for elt in self.data.iter() {
+                        try!(<u8 as Protocol>::proto_encode(elt, dst));
```
Comment (fenhl) [inline src/packet.rs]: Can we not use [`write_all`](http://doc.rust-lang.org/std/io/trait.Write.html#method.write_all) for this instead of a loop?

Comment (toqueteos) [inline src/packet.rs]: Yes! `write_all(&self.data)` should do!

**Comment by toqueteos** [comment]
On hold again, custom packets are still a bit picky.

**Comment by toqueteos** [comment]
Updated for #28 :tada:

I had to add `util::ReadExactExt` inside clientbound and serverbound modules.

File: src/packet.rs [hunk 2]
```diff
@@ -459,7 +459,21 @@ packets! {
             // 0x3c => UpdateScore { score_name: String, action: ScoreAction }
             0x3d => DisplayScoreboard { position: i8, score_name: String }
             // 0x3e => UpdateTeam { team_name: String, action: TeamAction }
-            // 0x3f => PluginMessage { channel: String, data: Vec<u8>; impl Packet for PluginMessage { ... } } // PROBLEM: length of `data` comes from packet length
+            0x3f => PluginMessage { channel: String, data: Vec<u8>;
+                impl Packet for PluginMessage {
+                    fn encode(&self, mut dst: &mut Write) -> io::Result<()> {
+                        try!(<String as Protocol>::proto_encode(self.channel, dst));
```
Comment (fenhl) [inline src/packet.rs]: This needs to be `&self.channel`.

File: src/packet.rs [hunk 3]
```diff
@@ -459,7 +459,21 @@ packets! {
...
+                        try!(dst.write_all(&self.data));
+                        Ok(())
+                    }
+                    fn decode(mut src: &mut Read, len: usize) -> io::Result<$name> {
```
Comment (fenhl) [inline src/packet.rs]: This needs to be `io::Result<PluginMessage>`.

**Comment by fenhl** [comment]
:+1: looks good to merge.

**Pull request merged**